### PR TITLE
neovim: add 'autowrapRuntimeDeps' to append

### LINKED ISSUE #352738: collect and add to PATH runtimeinputs of neovim plugin dependencies
## Issue description

There has been talk to better handle runtime inputs for nixpkgs.
Now that the wrapper sees the plugins directly it's easier to collect plugin runtime dependencies. 
I am not sure how other ecosystems do it (`runtimeDependencies` ? passthru.runtimeDeps ? ) but we could go through each plugin dependencies and add them to PATH in the wrapper.

For reference: 
https://github.com/NixOS/nixpkgs/issues/347093
https://github.com/NixOS/nixpkgs/pull/208029

cc interested folks @khaneliman @GaetanLepage @figsoda @PerchunPak @mrcjkb 

This should probably wait for after the nixos 24.11 release to give us more time to experiment but a proposition could be made beforehand.

### PR DESCRIPTION
**Pull request by teto**
Fix https://github.com/NixOS/nixpkgs/issues/352738

TLDR, adds runtime dependencies of vim plugins to the neovim wrapper PATH

I based it upon https://github.com/NixOS/nixpkgs/pull/356271 which hopefully will get merged

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/applications/editors/neovim/wrapper.nix [hunk 1]
```diff
@@ -24,6 +24,10 @@ let
     # to work with nix.
     # if true, the wrapper automatically appends those snippets when necessary
     , autoconfigure ? false
+
+    # append to PATH runtime deps of plugins
+    , autowrapRuntimeDeps ? false
```
Comment (PerchunPak) [inline pkgs/applications/editors/neovim/wrapper.nix:29]: Why having it opt-in? We already apply a wrapper to each plugin, no?

Comment (teto) [reply pkgs/applications/editors/neovim/wrapper.nix:29]: I thought we might have it opt-in at first to get feedback,to update the plugins with their runtimeDeps as well. And in a second step we would make it opt-out. We coul opt-in straightaway as well, I dont mind.

Comment (teto) [reply pkgs/applications/editors/neovim/wrapper.nix:29]: one issue with the opt-out is that it forces us to update the doc in a way that works  with the legacy and new wrapper, which should be done but I am too lazy to do yet :'(  I would rather document the new wrapper in another PR

**Comment by teto** [comment]
the test `nix-build -A neovim.tests.autowrap_runtime_deps` now passes.
the neovim used in the test `nix-build -A neovim.tests.nvim_with_runtimeDeps ` correctly wraps xxd when installing the plugin       `pkgs.vimPlugins.hex-nvim` with `autowrapRuntimeDeps = true;`.

Feel free to test it out thought it wont do much without updated plugins. The candidates could be:
- sniprun
- every lsp plugin, schemastore to include yamlls for instance (but can it work with others).

Feel free to complete the list.

File: pkgs/applications/editors/neovim/wrapper.nix (hunk 1 above)
Comment (PerchunPak) [reply pkgs/applications/editors/neovim/wrapper.nix:29]: I don't see a reason for changing plugins to use `runtimeDeps` in a subsequent PR. We can do it here to tackle problems before merging the feature. Unless you don't want to do this right now

But yeah, we can keep it opt-in for now, though I doubt someone would use opt-in undocumented features outside of [@]NixOS/neovim team

**Comment by khaneliman** [comment]
> the test `nix-build -A neovim.tests.autowrap_runtime_deps` now passes. the neovim used in the test `nix-build -A neovim.tests.nvim_with_runtimeDeps ` correctly wraps xxd when installing the plugin `pkgs.vimPlugins.hex-nvim` with `autowrapRuntimeDeps = true;`.
> 
> Feel free to test it out thought it wont do much without updated plugins. The candidates could be:
> 
>     * sniprun
> 
>     * every lsp plugin, schemastore to include yamlls for instance (but can it work with others).
> 
> 
> Feel free to complete the list.

Did you want us to add runtimeDeps to the packages in this PR or just add notes about other possible updates?

**Comment by teto** [comment]
Until we coordinate with vim such that they do the same as us, we should use autowrapRuntimeDeps for neovim plugins exclusively OR to add software we dont patch already, like the LSP servers. 

We can unpatch plugins later  if nixpkgs vim maintainers agree, discussion at: https://github.com/NixOS/nixpkgs/issues/352738#issuecomment-2547064119

I am tempted to merge under these conditions what do you think ?

**Comment by khaneliman** [comment]
> Until we coordinate with vim such that they do the same as us, we should use autowrapRuntimeDeps for neovim plugins exclusively OR to add software we dont patch already, like the LSP servers. 
> 
> 
> 
> We can unpatch plugins later  if nixpkgs vim maintainers agree, discussion at: https://github.com/NixOS/nixpkgs/issues/352738#issuecomment-2547064119
> 
> 
> 
> I am tempted to merge under these conditions what do you think ?

I think it's an improvement to allow wrapping runtime deps instead of patching. Even if we don't handle all occurrences immediately.

**Comment by teto** [comment]
let's add it to one or two plugins, if that works, we set it to true and mention  it in doc + 25.04 changelog.
